### ISSUE TITLE
Projection from mesh to rgb images incorrect

### ISSUE THREAD

**Issue by miguelriemoliveira** [lvonasek/tango#25]
Hi @lvonasek ,

as I told you I am back working on this. I am having some problems and wanted to get your opinion. So please, if you have some time read through this long (sorry) text and tell me what you think I am doing wrong.

My first task was to create a python script (that's my programming language) that would read an open constructor dataset, in particular:

1. The obj file in which the mesh is stored (given in your dataset folder, I am using the obj inside the dataset folder, I assume it is the non-simplified mesh)
2. The set of jpg rgb image files and corresponding txt files indicating the camera pose for each image (given in your dataset folder, in #20 you told me how they work)
3. The rgb camera's intrinsic parameters (I was not sure here, see bellow)
4. The per image ply or depth map (did not get this far yet)

I implemented the code reading all this stuff. As far as I can tell it should be working. To verify, what I do is to project the triangles in the mesh (1) to all the images described in (2). 

To get a better feeling, I select only a few triangles on the mesh which belong to an object (i.e. a door or a book), that is, I create a new mesh with meshlab by erasing all other faces,  and see if those triangles are well projected.

The odd thing is that they seem to be correctly projected for some images and incorrectly projected for others ... here's an example 

![screenshot from 2018-05-29 10-50-46](https://user-images.githubusercontent.com/11350998/40651673-32f09a12-632e-11e8-9c8b-71d6f25bf544.png)

... the object is the red book. For some images, e.g. 5,8,14,19,20 the projection seems ok. For others, e.g. 28,32,38 the triangles are just not well aligned with the book.

I see some reasons that could explain why this is not working, can you tell me your opinion?

a. There is something wrong with the timestamps -> It could be that, for some images the sync between the time of capturing the image and the time of getting the camera pose would fail. However, since it is the images which are close that are always wrong, it does not seem like a random unaligned timestamp problem.
b. I am doing something wrong with the projection -> It is very likely but I don's understand why the projection works for some images and not others ... In any case, my next step should be to  try to implement the projection as explained in #19 
c. I am getting the wrong intrinsics -> long explanation, see bellow. 

As for c. what I did was to use an app called "TangoROSStreamer" to publish camera_info_messages and then used it to create a yaml configuration file which I read. The content of the file is shown bellow. 

```yaml
image_width: 1920
image_height: 1080
camera_name: camera_color
camera_matrix:
  rows: 3
  cols: 3
  data: [1484.120151053405, 0.0, 982.1869163422028, 0.0, 1483.296808359715, 543.0569322723705, 0.0, 0.0, 1.0]
distortion_model: rational_polynomial
distortion_coefficients:
  rows: 1
  cols: 5
  data: [0.09370951394834764, -0.1640721737187069, 0.0, 0.0, 0.1224192868301668]
rectification_matrix:
  rows: 3
  cols: 3
  data: [1, 0, 0, 0, 1, 0, 0, 0, 1]
projection_matrix:
  rows: 3
  cols: 4
  data: [1484.120151053405, 0.0, 982.1869163422028, 0.0, 0.0, 1483.296808359715, 543.0569322723705, 0.0, 0.0, 0.0, 1.0, 0.0]
```

This calibration is for a 1920x1080 image, which makes sense since the images you have in the dataset are this size. However, when I use the Google Constructor they have a calibration file for a 960x540 camera: 

```yaml
camera_id: TANGO_3DR_CAMERA_COLOR
camera_intrinsics:
  calibration_type: TANGO_3DR_CALIBRATION_POLYNOMIAL_2_PARAMETERS
  width: 960
  height: 540
  model: [742.060075526702, 741.648404179857, 490.843458171101, 271.278466136185]
  distortion: [0.0937095139483476, -0.164072173718707]
camera_extrinsics:
  orientation: [-0.706785140012307, 0.707372138513566, 0.00647751405612829, 0.00612089243330192]
  translation: [-0.0178448539838816, 0.00721245490079159, -0.00481233577874809]
```

... my guess is that they use the camera at half the resolution. Do these values make sense for you ... do you have similar values?

Thanks in advance,
Miguel

**Comment by lvonasek** [comment]
Hi,

1. yes it is not simplified mesh

Google Constructor is saving camera frames as PNG (or PNM I am not sure). They had to decrease resolution because their dataset was getting very big (e.g. 1GB/1minute). The calibration from TangoROSStreamer looks for me familiar.

What I think it can be mistake is mismatching ortho-perspective projection. If you just transform vertex by matrix you get ortho matrix but the camera image is perspective.

After transfering vertex v into image coordinates (vec4 uv = world2uv * vec4(x,y,z,1)) you have to do perspective dividion which is normally uv /= abs(uv.w) but in your transform was no projection matrix, also you have to convert it this way uv /= abs(uv.z * uv.w) - just like it was discribed in https://github.com/lvonasek/tango/issues/19 (I am not sure if uv.z should not be without abs to have correctly sign of z coordinate).

I guess that some images were projected correctly because the distance from camera to point was 1 meter and then abs(1.0 * uv.w) == abs(uv.z * uv.w).

Hopefully this helps

Lubos

**Comment by miguelriemoliveira** [comment]
Hi @lvonasek ,

thanks for the clarifications. I will see if I can make it work and get back to you.

Regards,
Miguel

**Comment by miguelriemoliveira** [comment]
Hi @lvonasek ,

I am still struggling with making this work. I have implemented as you suggest but I still don't get good results. 

a couple of questions just to make sure I am not doing something wrong?

**1. Camera poses**

In #20 you told me that from a txt file like this

```
3
0.768300 -0.006883 -0.640053 0.000000
0.020602 -0.999158 0.035474 0.000000
-0.639758 -0.040441 -0.767511 0.000000
-0.293721 0.007573 0.122499 1.000000
-0.763887 0.005422 0.645327 0.000000
-0.025907 0.998901 -0.039059 0.000000
-0.644830 -0.046555 -0.762907 0.000000
-0.283288 0.007842 0.118334 1.000000
-0.029717 0.998833 -0.038075 0.000000
-0.760676 0.002113 0.649128 0.000000
0.648451 0.048252 0.759725 0.000000
-0.254234 0.000720 0.096176 1.000000
248.445531
```

This is rgb camera transform 
```
0.768300 -0.006883 -0.640053 0.000000
0.020602 -0.999158 0.035474 0.000000
-0.639758 -0.040441 -0.767511 0.000000
-0.293721 0.007573 0.122499 1.000000
```

This transform is from the global reference frame to the rgb camera's pose when the image was taken, correct?

The mesh (the obj file) you produce is also defined in this global reference frame? So if I need to project a triangle vertex from the mesh with coordinates given in the obj file I first need to put these obj coordinates into camera coordinates correct? That means multiplying the point coordinates by the global frame to the camera frame transformation, right? Only after I multiply by the world2uv matrix correct?

**2. Are these poses before or after the optimization**

Because if they are before that could explain the small projection errors ...

**Comment by miguelriemoliveira** [comment]
Hi again,
just to add to what I said. I computed the projection (pixel or texture coordinates) in 3 different ways. Always the same result. So now I am starting to wonder if there is another problem somewhere ...

**Comment by lvonasek** [comment]
I have no idea, I tried myself to project 3d vertices into image and it is working for me:
https://youtu.be/cLk5rxXKbaE
You can take a look into code, I put it into this repository.

**Comment by miguelriemoliveira** [comment]
Hi Lubos,

first of all sorry for the late reply. I thought that, with the end of my classes I would have more time, but I have several M.Sc. students asking me to revise their thesis proposals and this is taking a lot of my time.

Anyway, I saw your tutorial video and its really helpful. Thank you for that. I also downloaded, compiled and tried out your dataset_viewer tool, which is quite nice.

I have used your dataset viewer tool to inspect my "red book" dataset and compared it with my projection of triangles to the image. Here are the images:

In your case, you project the image information to the 3D mesh. The top left corner of the book is not well aligned, which is why you map red color to portions of the table surface.

![dataset_viewer](https://user-images.githubusercontent.com/11350998/41035559-eef7a3ca-6984-11e8-8328-c9d5763cdde9.png)

In my case, I do the opposite - I project the 3D meshes vertices to the image. We can also see the mesh (I used only the triangles which correspond to the book) miss-aligned with the book. This would result, as in your case, in painting a part of the table with red color from the top left corner of the book.

![miguels_viewer](https://user-images.githubusercontent.com/11350998/41035567-f6e1f0f4-6984-11e8-813f-3758d780bba0.png)

**So my conclusion (please tell me if you agree) is that my projection is being well done, since It returns the same results as in your case**. I am closing this issue. Thanks for all your support.

Just a side note: from what I understood, you are not using the compensation of distortion parameters of the camera's lens.  See [opencv](https://docs.opencv.org/2.4/modules/calib3d/doc/camera_calibration_and_3d_reconstruction.html#), the x'' and y'' equations. 

If you are saving the undistored rgb images (see [this](https://docs.opencv.org/2.4/modules/imgproc/doc/geometric_transformations.html#void%20undistort(InputArray%20src,%20OutputArray%20dst,%20InputArray%20cameraMatrix,%20InputArray%20distCoeffs,%20InputArray%20newCameraMatrix))) then you don't need to do the distortion correction while doing the projection. However, if you are just saving the raw rgb images, then you should do the distortion correction. 

I am not sure if the tango rgb camera has a lot of distortion (the numbers are bellow) so this effect could be minimal or not.

```
distortion_model: rational_polynomial
distortion_coefficients:
  rows: 1
  cols: 5
  data: [0.09370951394834764, -0.1640721737187069, 0.0, 0.0, 0.1224192868301668]
```

Can you confirm which images you are saving (raw or undistorted)?

If you want I can do a study (in a new issue) of the impact of using these distortion values versus not using them, just to assess if they are really important. That's the least I could do after you helped me so much.

In any case I will close this issue.

Regards,
Miguel

**Comment by lvonasek** [comment]
Hi Miguel,

I am using raw images. Distortion is applied during texturing (postprocess), in my dataset viewerI ignore it.

I forgot to tell you one important detail - the obj model is little bit simplified because there is no standart how to save per vertex colors and generating texture does not work with too detailed geometry.

If you want to do a research connected to 3D scanning and improving the scan quality, I have one problem:
if I scan flat surface with a grid pattern (floor in my kitchen) then the photos are not applied correctly - the motion tracking generates some small mistake which makes in the end from the grid pattern some kind of waves.
The first attempt to solve this was adding into settings "use sharp photos only" (because I thought the source of the problem are just blured photos but it is just one part of the problem).

The goal of the research would be to correct the camera matrices in the dataset, also in the end the texturing should result completely. I would provide all you ask for (if I am able to deliver it).

I am developing last months an ARcore version of the scanner:
https://plus.google.com/106761694899705891407/posts/V35843UquAz
and there is the problem with motion tracking much bigger.

Regards

Luboš

**Comment by miguelriemoliveira** [comment]
Hi Lubos,

to optimize the pose of each camera in the world so that we can align the images better is, from what I know, a process called Multi-view stere (MVS), structure from motion (SFM) or Bundle adjustment (BA).

From your description of the tiled floor I think I saw this happening also.

![screenshot from 2018-06-06 15-36-48](https://user-images.githubusercontent.com/11350998/41045395-4062079e-69a0-11e8-843c-16916de41cb1.png)

![screenshot from 2018-06-06 15-41-13](https://user-images.githubusercontent.com/11350998/41045396-4086c44e-69a0-11e8-92e7-e2225c27797a.png)

My problem is that I have a feeling that these waves in tiled surfaces are due to an error in RGBD sensors (in the images above the top is taken with a lenovo phab2pro, the second with a kinect v2) so I am not sure that this will be solved by changing the pose of the cameras. I think it is the actual 3D raw measurements from the sensor that are incorrect.

In any case I have a student working on BA for another application so I can safely say that this is definitely on my line of research. However, to be honest with you, my goal now in these next couple of months without classes is to try to contribute to the texture mapping algorithms with the idea I explained before assuming the geometric registration as it is given. In this I expect to have some results in the next few months and let you know about them.

Actually I though you were already doing some kind of optimization like BA already when the app said "optimizing model".

Regards,
Miguel

**Comment by lvonasek** [comment]
The app is doing this during postprocessing:
http://ed.ilogues.com/2016/02/02/3d-hdr-scene-capture (it is a part of Tango3DR library)

About incorrect model data you are right. For example here: https://skfb.ly/6yYzE - it is visible that floor is from two parts. When the motion tracking "jumps" then the frames are in slightly different coordinate system. The model geometry is not correct but I would like to have sharp textures at least on the flat surfaces. I can do a modification which would process just part of photo if the photo is on more surfaces.

**Comment by miguelriemoliveira** [comment]
Hi Lubos,

The first link is great. I am learnign a lot from you.

Concerning you suggested modification: I do not understand how you propose to do it or why.

Because if the problem is that the 3D mesh is wrong (with waves) then you may find some criteria for discarding a photo or parts of a photo. But in the end you will always have to use one of the photos in your dataset.

Because the 3D mesh is incorrect (does not match the reality),  think that the projection from a photo (any photo) will always have an incorrect projection, and changing the camera pose will not solve this since the problem is the wrongly captured 3D structure.

You say you would like to have sharp photos at least in the flat surfaces. Ok but do you have problems in these regions? Problems related to the waves cause by tiled surfaces?

When you mention a jump ... do you mean this region?

![lubos1](https://user-images.githubusercontent.com/11350998/41048714-4324210e-69a7-11e8-8b9b-8a009913f897.png)

I am sorry, I may be missing something but I don't understand your idea for the modification you suggest.

Regards,
Miguel

**Comment by lvonasek** [comment]
Yeah the jump is exactly on the region you marked. There is also second mismatch on the wooden frame and then you see that whole part would need few degress to be rotated.

The problem I would like to solve is on this flat surface but with grid texture:
![screenshot_20180606_185958](https://user-images.githubusercontent.com/6472545/41053571-417e495a-69bc-11e8-8ec5-65532d9d4774.png)

This small part you see on the picture is created by e.g. 50 photos, also I can theoretically simulate texturing and use only the matching majority of photos but then I have untextured parts of models and I need to fix it somehow... Maybe it would be better to modify photos, not poses... as you said the mesh is not matching reality and it can be unsolvable to make it all matching just with pose modification.

**Comment by miguelriemoliveira** [comment]
Hi Lubos,

One question: the raw measured data, i.e., the point coulds of these tiled surfaces already have these wave like patterns correct? Another question: do all the images that view the tiled surface show this error? Or is it that only some images, for example those which have the camera very oblique (almost paralell) to the surface show this error?


One idea: If there is a way to detect these tiled surfaces and the images belonging to cameras positioned in a way that will measure them  incorrectly, perhaps then you could say that the 3D data from these images is not used for the reconstruction. This would result in a reconstructed surface without the waves.

The problem is that this way I think you would have to do your own reconstruction and I am not sure you will not have to open the tango reconstruction blackbox ... So I would say that trying this out would depend on how much control you have of the reconstruction process.

I hope it helps. Regards,
Miguel

**Comment by lvonasek** [comment]
Hi Miguel,

I appreciate that you are trying to help but I think you still did not get the problem. Maybe I should not call it waves but blured/doubled texture. The problem is that motion tracking is containing drift... in the end the flat surface which is in the reality 3x5m is in the model e.g. 3.05x4.9m. During applying the photos there must be a place where this 0.05x0.1m difference is visible (for example some captured photos/poses are matching the reality and some of them are matching the model). In the end you will get the result like I posted in the previous post.

One solution would be to correct poses and do again reconstruction (thats with Tango3DR possible but I do not want to do that because it would be too slow on mobile device).

The second solution is to keep model and "bend" the photos:
-apply all photos which are not colliding (by simulating textuing and testing if pixels are matching)
-modify photos which are colliding (by remapping the coordinates of the photos to make it match with texture)

Regards
Lubos

**Comment by miguelriemoliveira** [comment]
Hi Lubos,

Ok, so you say only the second solution is interesting and it is about bending the photos.

I think what is difficult for me to understand is how you can modify photos which are colliding. If it is as you say by remapping the coordinates of the photos to make it match with texture this is only about changing the pose of the camera, right?

So that's the same as changing the position in space of a rigid body. There is no bending. To me bending is about changing the shape of the photo, e.g. changing position of the pixels w.r.t. each other.

In summary, from what I can understand:

1. changing the pose of the photos (as in a rigid body transform manner) would not be sufficient to get a correct alignment
2. changing also the shape of the photos, i.e. bending the photos could work but I have no idea how this could be done.

I am sorry I am not being very helpful ...

Regards,
Miguel

**Comment by lvonasek** [comment]
Hi Miguel,

I was deeply thinking about it and it is not easy as there are more problems I want to solve. Tango3DR is computing average color from all applied colors and thats resulting "ghosts". I would like to try to use median colors (I tried to imagine results and it should in many cases help a lot. E.g. in the screenshot I posted - there would be always just "the most dark" line in my grid):
1) let Tango3DR library do the texturing
-by this I have the color calibration of model
2) reproject the model into photos and do the color calibration of photos
-by this I have color calibrated photos
3) for every pixel in texture(s) calculate "median" color (by projecting all photos on model)
-by this I have the corrected texture (hopefully)

I know it is not 100% solution but it is about improvement...

Have a nice weekend

Lubos

**Comment by miguelriemoliveira** [comment]
Hi Lubos,

I am sorry for the delay. I was out all weekend.

The median seems to me like a good idea. I cannot fully understand how you propose to do it (because I think this procedure must be done triangle (mesh face) wise, not pixel wise).

In any case the median could in fact solve the problem. At least it will definitely solve the problems of averaging. So I do think it is a good idea.

Regards,
Miguel

**Comment by lvonasek** [comment]
Hi Miguel,

the triangle is too big unit - if I have for an example a bottle on the surface then the meshing makes it much more smaller (in comparation with the reality). After Tango3DR texturing only small parts of triangles are deffected with the averaging.

My calendar is currently quite full, I think I will have time for it since August and then it can be on September in the app.

Regards,
Lubos

**Comment by miguelriemoliveira** [comment]
Hi Lubos,

OK. I am definitely interested on your work. If you think I can help (in my country we say: help a lot those which do not get in the way! ) please keep me posted on the developments.

I will do the same with my part.

Regards,
Miguel

**Comment by lvonasek** [comment]
I will give you know as soon as I do some progress but as I said, it can take some time.

**Comment by lvonasek** [comment]
Hi Miguel,

I commited here the medianer_test which is an inital work on new texturing. There is still no color calibration, I started with the comparation of possible texturing methods.

This is how Tango SDK textures the model:
![tango](https://user-images.githubusercontent.com/6472545/42707448-86ca0766-86da-11e8-820d-00293e5b80b6.png)

I tried to write the same algorithm (color averaging):
![average](https://user-images.githubusercontent.com/6472545/42707479-a221f67c-86da-11e8-8d5b-355403ee6fa8.png)

It is visible that on the left bottom the Tango version is better (I guess that they are doing the averaging using increasing method e.g. color = color * 0.9 + newColor * 0.1, also the first frame has always priority)

---

I also tried to use always the nearest photo, which makes clear how big the error of the motion tracking is:
![nearest](https://user-images.githubusercontent.com/6472545/42707614-0e28e1c8-86db-11e8-9e58-edb5ebc464d1.png)

---

The median method is not easy to do, currently it is extremely slow and it is using too much memory. It is not easy to do a median without storing all informations. I also did median extra for every color channel (thats causing some color artifacts) but for the first iteration it is not bad:
![median](https://user-images.githubusercontent.com/6472545/42707864-cadec88c-86db-11e8-8d0b-592e9794e4fb.png)

Regards
Lubos

**Comment by miguelriemoliveira** [comment]
Hi @lvonasek ,

sorry for my delayed response. You caught me on a week of holidays. I try to disconnect as much as I can, otherwise I do not rest adequately.

Now to your message:

**About Tango**
If your guess is right, "I guess that they are doing the averaging using increasing method e.g. color = color * 0.9 + newColor * 0.1, also the first frame has always priority", It is amazing how simple the algorithm is (just a weighted average). I think there is a lot of space to improve here.

**About Your Color Averaging**
Results are very similar to tango's which suggests that your guess about the tangos algorithm is correct. Averaging I think will always have problems, since it is the averaging method that creates the phantom artifacts.

**About the nearest photo**
Yeah, I was expecting a better performance from the tango's motion tracking. If the geometric registration is that bad, I am amazed of the quality of the textured 3d models tango (and you) achieve.

**About the median**
I agree results are encouraging. The left part of the image looks much better. I am not sure about the part on the right though ... what do you plan to do next?

I am not sure if you know this already but I found some stuff online about the moving median
http://code.activestate.com/recipes/576930/
https://rhettinger.wordpress.com/tag/running-median/
perhaps they help to make the algorithm faster

**Comment by lvonasek** [comment]
Hi Miguel,

you really do not have to sorry for being offline. It is completely understandable.

The more I am thinking about it then more I would like to go into "nearest" solution. With median I will never reach a photorealistic textures. The next steps will be:
1) color calibration (then I can do color comparations much easier)
2) I will group the photos which are matching together (I expected about 5 groups in mine test dataset)
3) texturize it where the priority of frames will be by group ID (the first one will be the biggest)
4) TODO:thinking - do some hack on places where the slightly different coordinate systems meets (current idea fadeout)

**Comment by miguelriemoliveira** [comment]
Hi Lubos,

Thanks for the tip.

I planned to do some work during august but so for I could not start
anything. Hopefully in the second half of august I will start full throttle
:).

Keep in touch,

Miguel
